### PR TITLE
Upgrading to v2.4.0

### PR DESCRIPTION
**Pull request by metinavci**

### CODE REVIEW

File: client/src/main/java/com/assemblypayments/spi/model/TerminalStatusRequest.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import com.assemblypayments.spi.util.Events;
+import com.assemblypayments.spi.util.RequestIdHelper;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class TerminalStatusRequest {
+
+    public TerminalStatusRequest() { }
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/model/TerminalStatusRequest.java]: Redundant.

File: client/src/main/java/com/assemblypayments/spi/Spi.java [hunk 2]
```diff
@@ -1205,6 +1230,57 @@ private void stopTransactionMonitoring() {
...
+                this.hasSetInfo = true;
+                LOG.info("Setting POS info successful");
+            } else {
+                LOG.warn("Setting POS info failed: reason=" + response.getErrorReason() + ", detail=" + response.getErrorDetail());
+            }
+        }
+    }
+
+    private void handlePrintingResponse(@NotNull Message m) {
+        assert printingResponseDelegate != null;
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/Spi.java]: These asserts do nothing and your `printingResponse` doesn't have a `@NotNull` contract anyway. If it's intended to be not null, then annotate the arguments in delegate methods as well and remove these asserts (for all three).

File: client/src/main/java/com/assemblypayments/spi/model/TerminalStatusRequest.java (hunk 1 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/model/TerminalStatusRequest.java]: Fixed

File: client/src/main/java/com/assemblypayments/spi/Spi.java (hunk 2 above)
Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/Spi.java]: Removed asserts

File: client/src/main/java/com/assemblypayments/spi/Spi.java [hunk 3]
```diff
@@ -1603,39 +1682,15 @@ public synchronized GeneralSecurityException getCause() {
...
-            if (response.isSuccess()) return;
-
-            LOG.warn("Failed to cancel transaction: reason=" + response.getErrorReason() + ", detail=" + response.getErrorDetail());
-
-            txState.cancelFailed("Failed to cancel transaction: " + response.getErrorDetail() + ". Check EFTPOS.");
-
-            txFlowStateChanged();
-        }
+    public interface PrintingResponseDelegate {
+        void printingResponse(Message message);
```
Comment (mgouline) [inline client/src/main/java/com/assemblypayments/spi/Spi.java]: Still doesn't have corresponding `@NotNull` annotations, as I mentioned in the last review.

Comment (metinavci) [reply client/src/main/java/com/assemblypayments/spi/Spi.java]: Note: I coded these methods as previous in the Library. So, They should be fixed too. 
Removed @NotNull label for these 3 methods which I added.

**Pull request closed without merging**